### DISCUSSION TITLE
Do we support encoder/layout ?

### DISCUSSION THREAD

**Discussion by Alceatraz** [loki4j/loki-logback-appender#178] [General]
Hello, And:

As a skywalking demo, Dose loki4j support encoder/layout?

```xml
 <appender name="appender_sky" class="ch.qos.logback.core.ConsoleAppender">
    <encoder class="ch.qos.logback.core.encoder.LayoutWrappingEncoder">
      <layout class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.TraceIdPatternLogbackLayout">
        <pattern>%level [%tid][%thread/%logger{0}]%m%n</pattern>
      </layout>
    </encoder>
  </appender>
```

**Comment by nehaev** [comment]
Hi @Alceatraz,

No it's not supported, but maybe you can MDC instead.

**Reply by slauth** [reply]
Curious on this one… is there a strong reason why it's not possible to use arbitrary encoders?

In production I'm using `net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder` from [logstash-logback-encoder](https://github.com/logfellow/logstash-logback-encoder). I'd consider using `loki-logback-appender` locally only if it would support this encoder (because I want logs to look exactly the same as in production)…

**Reply by nehaev** [reply]
Hi @slauth, the things have changed since then. Starting from Loki4j v1.5.0 there is a custom layout support for messages. See [docs](https://loki4j.github.io/loki-logback-appender/docs/jsonlayout) for more details.

**Reply by slauth** [reply]
Thanks for the quick response. I did see this in the docs. However for me it would be important to use the exact same encoder that is being used in production…

**Reply by slauth** [reply]
Maybe to explain my use case a bit more: 

In production my application logs to `stdout` using the before mentioned `LoggingEventCompositeJsonEncoder` (logs are then automatically ingested to Loki by some infrastructure components). Now when developing locally, I'm currently using a `FileAppender` that uses the exact same encoder. Then, I use `promtail` to read this file and ingest to a Loki instance running locally (via Docker Compose). Would be nice if I wouldn't need the detour via file writing and reading but could directly push to Loki…

**Reply by nehaev** [reply]
You don't need the encoder. Instead, you can customize a message layout as it's outlined in the docs I've referred to. `LoggingEventCompositeJsonLayout` is identical to `LoggingEventCompositeJsonEncoder` in terms of configuration. I mean, no guarantees, but theoretically this should work.

It should look something like this:

```xml
<appender name="LOKI" class="com.github.loki4j.logback.Loki4jAppender">
    <format>
        ...
        <message class="net.logstash.logback.layout.LoggingEventCompositeJsonLayout">
            <!-- your logstash configuration goes here -->
        </message>
    </format>
    ...
</appender>
```

**Reply by slauth** [reply]
Ah okay so instead of re-useing the entire encoder I can re-use the same layout. Thanks!

I guess what confused me a bit here is the different namings (an XML element named `<message>` which sets a `Layout`).

**Comment by nehaev** [comment]
Let me correct my previous answer. 

Starting from Loki4j v1.5.0 there is a custom layout support for messages. See [JSON layout example](https://loki4j.github.io/loki-logback-appender/docs/jsonlayout) for more details.
